osu.Server.Spectator/Hubs/MultiplayerQueue.cs

```diff
@@ -0,0 +1,279 @@
+        public async Task AddItem(MultiplayerPlaylistItem item, MultiplayerRoomUser user)
+        {
+            if (room.Settings.QueueMode == QueueMode.HostOnly && !user.Equals(room.Host))
+                throw new NotHostException();
+
+            if (room.Settings.QueueMode != QueueMode.HostOnly && room.Playlist.Count(i => i.OwnerID == user.UserID && !i.Expired) >= PER_USER_LIMIT)
+                throw new InvalidStateException($"Can't enqueue more than {PER_USER_LIMIT} items at once.");
+
+            using (var db = dbFactory.GetInstance())
+            {
+                string? beatmapChecksum = await db.GetBeatmapChecksumAsync(item.BeatmapID);
+
+                if (beatmapChecksum == null)
+                    throw new InvalidStateException("Attempted to add a beatmap which does not exist online.");
+
+                if (item.BeatmapChecksum != beatmapChecksum)
+                    throw new InvalidStateException("Attempted to add a beatmap which has been modified.");
+
+                if (item.RulesetID < 0 || item.RulesetID > ILegacyRuleset.MAX_LEGACY_RULESET_ID)
+                    throw new InvalidStateException("Attempted to select an unsupported ruleset.");
+
+                ensureModsValid(item);
+
+                switch (room.Settings.QueueMode)
+                {
+                    case QueueMode.HostOnly:
+                        // In host-only mode, the current item is re-used.
+                        item.ID = CurrentItem.ID;
+                        item.OwnerID = CurrentItem.OwnerID;
+
+                        await db.UpdatePlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, item));
+                        room.Playlist[currentIndex] = item;
+
+                        await hub.OnPlaylistItemChanged(room, item);
+                        break;
+
+                    default:
+                        item.OwnerID = user.UserID;
+                        item.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, item));
+                        room.Playlist.Add(item);
+
+                        await hub.OnPlaylistItemAdded(room, item);
+
+                        // The current item can change as a result of an item being added. For example, if all items earlier in the queue were expired.
+                        await updateCurrentItem();
+                        break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Checks whether the given mods are compatible with the current playlist item's mods and ruleset.
+        /// </summary>
+        /// <param name="proposedMods">The mods to check against the current playlist item.</param>
+        /// <param name="validMods">The set of mods which _are_ valid.</param>
+        /// <returns>Whether all mods are valid for the current playlist item.</returns>
+        public bool ValidateMods(IEnumerable<APIMod> proposedMods, [NotNullWhen(false)] out IEnumerable<APIMod>? validMods)
```

Comment:
Looks like this only uses the `CurrentItem`, would this potentially sit better in the model class? Not sure if there would be uses osu!-side, but being in the queue implementation makes it seem like the validation will change per queue type.

Reply:
Also I'd probably call this `ValidateUserMods` if that's the intention, since there's a second different validation method direction below this which is used in different scenarios.

Reply:
By model you mean `MultiplayerRoom`?

Hmm... Potentially. The model only contains `PlaylistItemId` so no real connection to the model itself, but there's two ways around that:
- Replace `MultiplayerRoomSettings.PlaylistItemId` with a `MultiplayerPlaylistItem`, denoting the current item.
- Do a manual lookup every time (`CurrentItem => Playlist.First(item => item.ID == settings.PlaylistItemId)`.

Is that what you're envisioning?

Reply:
Perhaps in the future multiplayer could be updated so the host calls `ChangeSettings()` with the new settings for the current playlist item...? Rather than `AddItem()` to do the same (currently) or a separate `ChangeItem()` in the future.. Though we may still want other players to be able to change items they've enqueued, so not sure how helpful that addition is.

Reply:
I meant `MultiplayerPlaylistItem`

currently:
`match.Queue.ValidateMods(..)`
post:
`match.CurrentItem.ValidateMods(..)`

Reply:
Yeah, so should I be removing `MultiplayerRoomSettings.PlaylistItemId` then?

Reply:
Moved them to extension methods on `MultiplayerPlaylistItem`, currently only available on the server. Eventually will need to take in an `IRulesetStore` or something, at which point it can be moved to the osu!-side model.

Didn't move the current item out of the queue and into `ServerMultiplayerRoom` yet because that needs a bit more finessing - I was scared of potential O(n^2) operation when validating mods for all users in the room (e.g. upon host changing the playlist item).